Support chat, completions task output for text-generation pipeline

es94129 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/es94129/mlflow/pull/10872?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10872/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10872
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

This PR changes the `output` for the `text-generation` pipeline when the `transformers` model is logged with `task` being `llm/v1/completions` or `llm/v1/chat`.

- `save_model`: save `task` into `flavor_config` (with the key `inference_task`) and `metadata` (with the key `task`) if the argument `task` is `llm/v1/xxx`
- `predict`: if `flavor_config` has `inference_task` defined, the output would be wrapped into
```python
{
  "text": ... (or "messages": ...),
  "finish_reason": ...,
  "usage": {
    "prompt_tokens": ...,
    "completion_tokens": ...,
    "total_tokens": ...
  }
}
```

(Other `transformers` pipelines, input formats, stopping criteria etc are not implemented in this PR.)

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

es94129 committed:
Enable task type to be in llm/v1/completions or llm/v1/chat

Signed-off-by: Ying Chen <ying.chen@databricks.com>

es94129 committed:
Pass inference task type to _TransformersWrapper

Signed-off-by: Ying Chen <ying.chen@databricks.com>

es94129 committed:
Output dict for text-generation pipe w. task

Signed-off-by: Ying Chen <ying.chen@databricks.com>

B-Step62 commented on mlflow/transformers/__init__.py:
```diff
@@ -459,10 +469,12 @@ def save_model(
     code_dir_subpath = _validate_and_copy_code_paths(code_paths, str(path))
 
-    resolved_task = _get_or_infer_task_type(transformers_model, task)
+    resolved_transformers_task = _get_or_infer_transformers_task_type(transformers_model, task)
```
Imo this method is a bit hard to follow, mainly because the `task` remains ambiguous i.e. whether it is transformers or inference task. Some method implies it handles transformer task such as `_validate_transformers_task_type`, but it actually inputs inference task as well.

To make things simple, I think we should rename it to `transformer_task` or `inference_task` as early as possible, and handle them in separate method, instead of having single method input `task`.

For example:
```
transformers_task, inference_task = _get_or_infer_task_type(transformers_model, task)

_validate_transformers_task_type(transformers_task)
_validate_inference_task_type(inference_task)

...

def _get_or_infer_task_type(transformers_model, task):
    if not task:
        transformers_task = _infer_transformers_task(task)
        inference_task = _get_inference_task(task, model)
    elif _is_transformer_task(task, ...):
        transformers_task = task
        inference_task = _get_inference_task(task, model)
    else:
        transformers_task = _infer_transformers_task(model)
        inference_task = task

    return transformers_task, inferece_task
```

This makes the downstream logic simple as well, like we can do `if inference_task` rather than checking supported types for pipeline. WDYT?

es94129 replied:
I love that this makes the logic a lot simpler.
As a side note, do we want the `predict` function to always output a dictionary instead of a string even when `task` is not specified in the format of `llm/v1/*` and the input is in the normal Hugging Face format?

(Currently I haven't added the logic to infer the `inference_task` when it isn't provided as `task` yet.)

- If yes, it would change the behavior of `predict` quite a bit, but the implementation would be more straightforward
- If not, in addition to `inference_task`, we might also need to save whether `inference_task` is provided by user or inferred by us.

Either way, I will definitely need to refactor this a bit, but would love to hear your opinion of this.

B-Step62 replied:
Hmm actually I'm not sure if we can "infer" the inference task type? We don't know if user intend to use it as chat or completion. If that's not deterministic, we can't output dictionary but just follow original transformers output string. This effectively means that user has to set `llm/v1/*` task if they want the new interface.

In other words, the condition here changes from
> Behavior change when (a) task is defined as llm/v1/completions or llm/v1/chat, or
(b) data, params are in OpenAI format

to 

> Behavior change when (a) task is defined as llm/v1/completions or llm/v1/chat, ##and###
(b) data, params are in OpenAI format

I think this is fine, but if not, we might need some fundamental change in the design....

CC: @dbczumar

es94129 replied:
Got it! I think it would make sense to enforce it only when `task` is explicitly set as `llm/v1/*` -- it would be weird if people could pass arguments in OpenAI format but don't specify the `task` indeed.

B-Step62 commented on mlflow/transformers/__init__.py:
```diff
@@ -2001,7 +2038,13 @@ def _predict(self, data):
+            # If inference task is defined, return tensors instead of generated text
+            model_config = self.model_config.copy()
+            if self.inference_task:
+                model_config.update({"return_tensors": True})
```
Can we do this when saving model? It can be confusing that the model returns different thing as it is configured in the file. I guess it's possible because the inference task is declared when saving model.

es94129 replied:
Maybe the comment is misleading here, I intend to only return tensors internally and decode it as generated text + count the tokens within the `predict` function, so the behavior should be the same.

B-Step62 commented on mlflow/transformers/__init__.py:
```diff
@@ -2928,6 +2975,59 @@ def _format_prompt_template(self, input_data):
+        """
+        output_dicts = []
+        for input_data, output_tensor in zip(data, output):
+            # Get token usage
+            inputs = self.pipeline.tokenizer(
+                input_data,
+                return_tensors=self.pipeline.framework,
+                max_length=self.model_config.get("max_length", None),
+            )
+            prompt_tokens = inputs["input_ids"].shape[-1]
+            total_tokens = len(output_tensor)
+            completions_tokens = total_tokens - prompt_tokens
+
+            # Decode output from tensor
+            generated_text = self.pipeline.tokenizer.decode(
+                output_tensor,
+                skip_special_tokens=True,
+                clean_up_tokenization_spaces=True,
+            )
+
+            # Strip prompt from output
+            completions_text = generated_text[len(input_data) :].lstrip()
+
+            # Determine finish reason
+            finish_reason = "stop"
+            if total_tokens > self.model_config.get(
+                "max_length", float("inf")
+            ) or completions_tokens == self.model_config.get("max_new_tokens", float("inf")):
+                finish_reason = "length"
+
+            # Construct output dict
+            output_dict = {
+                "finish_reason": finish_reason,
+                "usage": {
+                    "prompt_tokens": prompt_tokens,
+                    "completion_tokens": len(output_tensor) - prompt_tokens,
+                    "total_tokens": len(output_tensor),
+                },
+            }
```
Can you make these steps into single method like `get_token_usage`, `determine_finish_reason` (probably worth moving to separate file as `__init__.py` becomes way too big already...) and add unit tests for each method + entire postprocess?  This is complex logic so error prone when updating in the future.

B-Step62 commented on mlflow/transformers/__init__.py:
```diff
@@ -94,6 +94,8 @@
 _INFERENCE_CONFIG_BINARY_KEY = "inference_config.txt"
 _INSTANCE_TYPE_KEY = "instance_type"
+_INFERENCE_TASK_KEY = "inference_task"
+_METADATA_INFERENCE_TASK_KEY = "task"
```
Shouldn't we use `inference_task` for consistency?

es94129 replied:
We should, but Databricks Provisioned Throughput uses `metadata: { "task": ... }` for making optimizations, so IMO saving to the `task` field here would prevent the need from logging the model with both
```python
mlflow.transformers.save_model(
  ...,
  task="llm/v1/completions", # <- added in this PR
  metadata={"task": "llm/v1/completions"}, # <- currently required
)
```

B-Step62 replied:
Makes sense, thanks! Could you add a comment describes that context for the future reference?

B-Step62 reviewed (commented):
Thanks for implementing the change! Left a few comments to make the logic simpler and robust to change.

es94129 committed:
Address comments

Signed-off-by: Ying Chen <ying.chen@databricks.com>

es94129 committed:
lint

Signed-off-by: Ying Chen <ying.chen@databricks.com>

B-Step62 commented on mlflow/transformers/__init__.py:
```diff
@@ -2928,6 +3005,19 @@ def _format_prompt_template(self, input_data):
 
+    def _postprocess_output_for_inference_task(self, data, output):
+        """
+        Wrap output data with usage information according to the MLflow inference task.
```
nit: (1) Could you add input/output example in doc string? (2) Maybe we can move this to `inference_utils.py`.

B-Step62 commented on mlflow/transformers/inference_utils.py:
```diff
@@ -0,0 +1,68 @@
+_INFERENCE_TASK_CHAT = "llm/v1/chat"
+
+
+def _get_output_and_usage_from_tensor(
```
nit: In Python conventionally, a function prefixed with an underscore `_` means a *private* function that is only used inside the module. This one is called from the other module so we shouldn't mark it with an underscore. Nothing prevents ppl to call *private* function outside the module so it's more like annotation. Also we know many files in mlflow violates this.. but it's good to keep new things clean :\

B-Step62 commented on mlflow/transformers/inference_utils.py:
```diff
@@ -0,0 +1,68 @@
+    prompt: str, output_tensor: List[int], pipeline, model_config, inference_task
+):
+    """
+    Decode the output tensor and return the output text and usage information as a dictionary.
```
nit: could you add a mention that this is to make output OpenAI compatible format.

B-Step62 commented on tests/transformers/test_transformers_inference_utils.py:
```diff
@@ -0,0 +1,51 @@
+)
+
+
+def test_output_dict_for_completions(text_generation_pipeline):
```
nit: Can we mock tokenizer to simple `lambda s: s.spilt(" ")` to make tests faster and simpler? Transformers test is super slow already.....

es94129 replied:
Got it, I just mocked all the return values of the tokenizer functions in the `test_transformers_llm_inference_utils.py` file

B-Step62 replied:
It still loads `text_generation_pipeline` fixture which takes time. I think you can do
```
class DummyTokenizer:
     input_ids = torch.tensor([1, 2, 3])

     def decode(self, s, ...):
        return s.split(" ")
...

pipeline = mock.MagicMock()
pipeline.tokenizer.return_value = DummyTokenizer()
```

es94129 replied:
Makes sense, thanks for pointing this out! Indeed the fixture loads models which is not efficient nor necessary in this case.

B-Step62 commented on mlflow/transformers/__init__.py:
```diff
@@ -94,6 +100,10 @@
 _IMAGE_PROCESSOR_TYPE_KEY = "image_processor_type"
 _INFERENCE_CONFIG_BINARY_KEY = "inference_config.txt"
 _INSTANCE_TYPE_KEY = "instance_type"
+_INFERENCE_TASK_KEY = "inference_task"
```
nit: Could you move these new constants incl. supported pipeline types to `inference_utils.py`, for better segregation?

B-Step62 reviewed (commented):
Added a few minot comments, but overall looks good to me!

Btw what do you think about renaming the task to `llm_inference_task`? Imo `inference_task` sounds a bit too general as every other task types also does "inference". It might be confusing for ppl without context.

We used `llm_` notion in the scoring server already ([example](https://github.com/mlflow/mlflow/blob/65c6b8c8dcdd9067b220913b3f019211e627028f/mlflow/pyfunc/scoring_server/__init__.py#L78)), so it would be more consistent?

es94129 committed:
Address comments

Signed-off-by: Ying Chen <ying.chen@databricks.com>

es94129 commented:
Thanks @B-Step62! I renamed the existing variables / file names to be `llm_inference_xxx` and moved the utility function and constants to the `llm_inference_utils.py` file.

B-Step62 reviewed (approved):
Thanks @es94129  for the quick turnaround! Could you also check if we can address [this comment](https://github.com/mlflow/mlflow/pull/10872#discussion_r1470403545)? Other than that looks good to me, let's get Corey's eyes on it and then we're good to go!

fyi: `build_doc` step failure is not related to this change so feel free to ignore (haru is working on the fix)

es94129 committed:
Mock all tokenizer usage in llm_inference_utils test

Signed-off-by: Ying Chen <ying.chen@databricks.com>

serena-ruan commented on mlflow/transformers/__init__.py:
```diff
@@ -1334,16 +1351,23 @@ def _extract_torch_dtype_if_set(pipeline):
+
+    if not task:
+        transformers_task = _infer_transformers_task_type(model)
+    elif _is_transformers_task_type(task):
+        transformers_task = task
     else:
-        task = _infer_transformers_task_type(model)
-    return task
+        transformers_task = _infer_transformers_task_type(model)
+        llm_inference_task = task
```
I feel like this function is super confusing... hard to read.
Could you update the docstring to make it more clear?
BTW could we move up _validate_llm_inference_task_type here?

serena-ruan commented on mlflow/transformers/__init__.py:
```diff
@@ -1837,13 +1891,17 @@ def _override_model_config(self, params):
             self.model_config.update(params)
 
-    def _validate_model_config_and_return_output(self, data):
+    def _validate_model_config_and_return_output(self, data, force_return_tensors=False):
```
```suggestion
    def _validate_model_config_and_return_output(self, data, return_tensors=False):
```

serena-ruan commented on mlflow/transformers/__init__.py:
```diff
@@ -1837,13 +1891,17 @@ def _override_model_config(self, params):
         import transformers
 
+        model_config = copy.deepcopy(self.model_config)
+        if force_return_tensors:
+            model_config["return_tensors"] = True
```
```suggestion
        model_config["return_tensors"] = return_tensors
```

serena-ruan commented on mlflow/transformers/llm_inference_utils.py:
```diff
@@ -0,0 +1,127 @@
+    if inference_task == _LLM_INFERENCE_TASK_COMPLETIONS:
+        output_dict["text"] = completions_text
+    elif inference_task == _LLM_INFERENCE_TASK_CHAT:
+        output_dict["messages"] = {"role": "assistant", "content": completions_text}
```
Is this role always `assistant`?

es94129 replied:
Yes, the LLM should always return the message in an `assistant` role.

serena-ruan commented on tests/transformers/test_transformers_model_export.py:
```diff
@@ -3862,3 +3875,66 @@ def _calculate_expected_size(path_or_dir):
+        )
+
+
+def test_text_generation_task_completions_predict_with_hf_params(
```
Could we add a test for serving as well?

es94129 replied:
Definitely, added `test_text_generation_task_completions_serve` now.
(As a future note, it still requires signature to be specified, and the request needs to be wrapped in `inputs` / `dataframe_split` etc, which would be addressed in followup PRs.)

serena-ruan commented on mlflow/transformers/llm_inference_utils.py:
```diff
@@ -0,0 +1,127 @@
+_LLM_INFERENCE_TASK_COMPLETIONS = "llm/v1/completions"
+_LLM_INFERENCE_TASK_CHAT = "llm/v1/chat"
+
+_SUPPORTED_LLM_INFERENCE_TASK_TYPES_BY_PIPELINE = {
```
Should this be `BY_PIPELINE_TASK` just to be clear?

es94129 committed:
Improve test mock and address comments

Signed-off-by: Ying Chen <ying.chen@databricks.com>

es94129 committed:
Add serve test

Signed-off-by: Ying Chen <ying.chen@databricks.com>

es94129 committed:
Test stability

Signed-off-by: Ying Chen <ying.chen@databricks.com>

es94129 committed:
Test stability

Signed-off-by: Ying Chen <ying.chen@databricks.com>

B-Step62 commented:
@es94129 merged the PR, let's do few follow-ups;

1. We haven't implement the input preprocessing for `llm/v1/chat` task right? As mentioned in the previous note, you can simply add a processing to map chat input to string using [ChatTemplate](https://huggingface.co/docs/transformers/chat_templating), 
2. It seems we need a quick tweak to make the new task type work with the new [prompt template](https://mlflow.org/docs/latest/llms/transformers/guide/index.html#saving-prompt-templates-with-transformer-pipelines) feature. When a prompt template is set, we set `return_full_text=False` for pipeline so it won't leak the system prompt (basically it prevents the pipeline to echo the user's input). The issue is that the parameter cannot be combined with `return_tensors` which we use for the new task type: `ValueError: 'return_full_text' is mutually exclusive with 'return_tensors'`. However, luckily this is not a big issue, indeed if we pass `return_tensors`, the generated tensor doesn't include the users' input but just new tokens. Hence, we can just override `return_full_text` to `True` when `return_tensors` is set [here](https://github.com/mlflow/mlflow/blob/master/mlflow/transformers/__init__.py#L1901-L1902).

Let me know if there is any unclear thing! Thanks!.

es94129 commented:
Thank you @B-Step62! I'll make sure to address those in follow-ups.
1. Absolutely, we haven't implemented any of the input preprocessing yet, including several main items which I will add in a follow-up PR:
   - Applying chat template for `llm/v1/chat`
   - Support OpenAI compatible parameters such as `max_tokens`, `stop`
   - Support dict input where the OpenAI compatible parameters are included in the `data` dict for the `predict` function
2.  Thanks for raising awareness to this and the pointers! Yes, we shall override `return_full_text` there, as we already have postprocessing util functions that remove the prompt text according to the logic of https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/src/transformers/pipelines/text_generation.py#L312
3. Another follow-up is to have model signature inferred when `task` is `llm/v1/xxx`.
